### PR TITLE
[MoE] Add node limited routing support

### PR DESCRIPTION
**Pull request by shuhuayu**
As titled, added node-limited routing support via two-layer routing.  First, group experts into `num_groups` groups, and experts in the same group should reside on the same node to utilize fast intra-node communication. Second, pick the `top_k_group` by the top 2 expert scores' sum in each group. Third, pick `top_k` experts within the selected `top_k_groups`. 

Reference: https://github.com/huggingface/transformers/blob/4c9fde2a2a3aece0bcf1be93f696e88297da9397/src/transformers/models/deepseek_v3/modeling_deepseek_v3.py#L212

Test on one node using DeepSeek V3 debug model with MoE arguments `num_experts=8,
            num_shared_experts=2,
            num_groups=4,
            top_k_group=2,
            top_k=3`.
Pasted Graphic

### CODE REVIEW

File: torchtitan/models/moe/moe.py [hunk 1]
```diff
@@ -26,7 +26,9 @@ class MoEArgs:
     route_scale: float = 1.0
     score_before_experts: bool = True
 
-    # token-choice
+    # token-choice with node limited routing support
+    num_groups: int | None = None  # must be a divisor of num_experts
+    top_k_group: int | None = None
```
Comment (wwwjn) [inline torchtitan/models/moe/moe.py]:
```suggestion
    top_k_groups: int | None = None
```

File: torchtitan/models/moe/moe.py [hunk 2]
```diff
@@ -225,6 +239,60 @@ def _debug_force_load_balance_routing(
...
+        # Mask out experts from non-selected groups
+        scores_for_choice = scores_for_choice.masked_fill(
+            ~score_mask.bool(), float("-inf")
+        )
+        # Select top_k experts from masked scores
+        selected_experts_indices = torch.topk(
+            scores_for_choice, k=self.top_k, dim=-1, sorted=False
+        )[1]
+        # Get actual scores (without bias) for the selected experts
+        top_scores = scores.gather(1, selected_experts_indices)
```
Comment (tianyu-l) [inline torchtitan/models/moe/moe.py]: let's unify this part with other two paths (even no expert bias we can do topk + gather)

File: torchtitan/models/moe/moe.py (hunk 1 above)
Comment (tianyu-l) [reply torchtitan/models/moe/moe.py]: set the default for 671B model according to https://github.com/deepseek-ai/DeepSeek-V3/blob/main/inference/configs/config_671B.json

Comment (tianyu-l) [inline torchtitan/models/moe/moe.py]: Are the names coming from some repo? I saw that deepseek repo calls them `n_expert_groups` and `n_limited_groups`.

I think we can combine the convention and call them `num_expert_groups` and `num_limited_groups`. WDYT?

File: torchtitan/models/moe/moe.py [hunk 3]
```diff
@@ -257,7 +325,17 @@ def forward(
...
         # NOTE: The expert_bias is only used for routing. The gating value
         #       top_scores is still derived from the original scores.
-        if expert_bias is not None:
+        if self.num_groups is not None:
+            assert (
+                self.top_k_group is not None
+            ), "top_k_group must be set when num_groups is set"
+            assert (
+                self.num_experts % self.num_groups == 0
+            ), f"num_experts ({self.num_experts}) must be divisible by num_groups ({self.num_groups})"
```
Comment (tianyu-l) [inline torchtitan/models/moe/moe.py]: Instead of doing assert, let's raise ValueError since they are more like error from users.

File: torchtitan/models/moe/moe.py (hunk 1 above)
Comment (tianyu-l) [reply torchtitan/models/moe/moe.py]: nit: let's put these two fields below `top_k` which is a more "important" arg.

**Comment by tianyu-l** [comment]
need to remove https://github.com/pytorch/torchtitan/blob/main/torchtitan/models/deepseek_v3/model/args.py#L70-L71

File: torchtitan/models/moe/moe.py (hunk 1 above)
Comment (shuhuayu) [reply torchtitan/models/moe/moe.py]: Sounds great! I changed to `num_expert_groups` and `num_limited_groups`, which have clearer meaning  from naming. The previous names are from huggingface's implementations.

File: torchtitan/models/moe/moe.py (hunk 2 above)
Comment (shuhuayu) [reply torchtitan/models/moe/moe.py]: Great suggestion! Merged into a unified `_node_limited_routing` method.

Comment (shuhuayu) [reply torchtitan/models/moe/moe.py]: Also changed to a smaller mask to save memory.

File: torchtitan/models/moe/moe.py [hunk 4]
```diff
@@ -225,6 +239,70 @@ def _debug_force_load_balance_routing(
...
+
+        selected_experts_indices = torch.topk(
+            scores_for_choice, k=self.top_k, dim=-1, sorted=False
+        )[1]
+
+        # NOTE: The expert_bias is only used for routing. The gating value
+        #       top_scores is still derived from the original scores.
+        top_scores = scores.gather(dim=1, index=selected_experts_indices)
+
+        return selected_experts_indices, top_scores
```
Comment (tianyu-l) [inline torchtitan/models/moe/moe.py]: this should stay outside node_limited_routing method. If you worry about naming, you could call it something like `_get_node_limited_routing_scores`

File: torchtitan/models/moe/moe.py [hunk 5]
```diff
@@ -180,9 +182,17 @@ class TokenChoiceTopKRouter(nn.Module):
     """This class implements token-choice routing. In token-choice top-K routing, each token is
         routed to top K experts based on the router scores.
 
+    Optionally supports node-limited (group-limited) routing where experts are divided into groups
+    (e.g., by node), and only num_limited_groups groups are considered before selecting top_k experts.
+    This reduces cross-node communication in distributed settings.
```
Comment (wwwjn) [inline torchtitan/models/moe/moe.py:187]: This is not true for NCCL native a2a, and only true with DeepEP?

File: torchtitan/models/moe/moe.py [hunk 6]
```diff
@@ -225,6 +239,70 @@ def _debug_force_load_balance_routing(
...
+                raise ValueError(
+                    f"experts_per_group ({experts_per_group}) must be >= 2"
+                )
+            scores_grouped = scores_for_choice.view(
+                -1, self.num_expert_groups, experts_per_group
+            )
+            group_scores = scores_grouped.topk(2, dim=-1)[0].sum(dim=-1)
+            group_idx = torch.topk(
+                group_scores, k=self.num_limited_groups, dim=-1, sorted=False
+            )[1]
```
Comment (wwwjn) [inline torchtitan/models/moe/moe.py]:
```suggestion
            _, group_idx = torch.topk(
                group_scores, k=self.num_limited_groups, dim=-1, sorted=False
            )
```

For readability, it's easy to forgot why we need to slice topk()'s result

File: torchtitan/models/moe/moe.py [hunk 7]
```diff
@@ -225,6 +239,70 @@ def _debug_force_load_balance_routing(
...
+            group_mask = torch.ones_like(group_scores, dtype=torch.bool)
+            group_mask.scatter_(1, group_idx, False)  # False = selected groups (keep)
+            # Mask out experts from non-selected groups
+            scores_for_choice = scores_grouped.masked_fill(
+                group_mask.unsqueeze(-1), float("-inf")
+            ).view(-1, self.num_experts)
+
+        selected_experts_indices = torch.topk(
+            scores_for_choice, k=self.top_k, dim=-1, sorted=False
+        )[1]
```
Comment (wwwjn) [inline torchtitan/models/moe/moe.py]:
```suggestion
        _, selected_experts_indices = torch.topk(
            scores_for_choice, k=self.top_k, dim=-1, sorted=False
        )
```

File: torchtitan/models/moe/moe.py [hunk 8]
```diff
@@ -225,6 +239,70 @@ def _debug_force_load_balance_routing(
...
+                )
+            experts_per_group = self.num_experts // self.num_expert_groups
+            if experts_per_group < 2:
+                raise ValueError(
+                    f"experts_per_group ({experts_per_group}) must be >= 2"
+                )
+            scores_grouped = scores_for_choice.view(
+                -1, self.num_expert_groups, experts_per_group
+            )
+            group_scores = scores_grouped.topk(2, dim=-1)[0].sum(dim=-1)
```
Comment (wwwjn) [inline torchtitan/models/moe/moe.py]:
```suggestion
            top2_scores_in_group, _ = scores_grouped.topk(2, dim=-1)
            group_scores = top2_scores_in_group.sum(dim=-1)
```

 wdyt?

File: torchtitan/models/moe/moe.py (hunk 5 above)
Comment (shuhuayu) [reply torchtitan/models/moe/moe.py:187]: I think this depends how the communication is implemented. The node-limited routing itself works the same for both all-to-all communication, i.e., enforcing communication across less nodes.

File: torchtitan/models/moe/moe.py (hunk 6 above)
Comment (shuhuayu) [reply torchtitan/models/moe/moe.py]: Incorporated, thx!

File: torchtitan/models/moe/moe.py (hunk 4 above)
Comment (shuhuayu) [reply torchtitan/models/moe/moe.py]: Make sense. Moved these lines into the `forward` method.

File: torchtitan/models/moe/moe.py [hunk 9]
```diff
@@ -225,6 +239,57 @@ def _debug_force_load_balance_routing(
...
+            scores: Router scores after sigmoid or softmax, shape (bs*slen, num_experts)
+            expert_bias: Optional bias for load balancing, shape (num_experts,)
+
+        Returns:
+            scores_for_choice: shape (bs*slen, num_experts)
+        """
+        scores_for_choice = scores if expert_bias is None else scores + expert_bias
+
+        # Apply node-limited routing mask if configured
+        if self.num_expert_groups is not None:
```
Comment (tianyu-l) [inline torchtitan/models/moe/moe.py]: this should happen outside, i.e.
```
if node_limited:
  _get_node_limited_routing_scores()
else:
  ...
```

Comment (shuhuayu) [reply torchtitan/models/moe/moe.py]: Nice suggestion! Moved this condition check to the `forward` method and now this method will process `scores_for_choice` only if node-limited routing is configured, which better aligns with its name.
